src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetVolumeInformationByHandle.cs

```diff
@@ -0,0 +1,76 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static partial class Kernel32
+    {
+        [LibraryImport(Libraries.Kernel32, EntryPoint = "GetVolumeInformationByHandleW", SetLastError = true, StringMarshalling = StringMarshalling.Utf16)]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        private static unsafe partial bool GetVolumeInformationByHandleW(
+            SafeFileHandle hFile,
+            char* volumeName,
+            int volumeNameBufLen,
+            int* volSerialNumber,
+            int* maxFileNameLen,
+            int* fileSystemFlags,
+            char* fileSystemName,
+            int fileSystemNameBufLen);
+
+        public static unsafe int GetVolumeInformationByHandle(
+            SafeFileHandle hFile,
+            out string? volumePath,
+            bool wantsVolumePath,
+            int* volSerialNumber,
+            int* maxFileNameLen,
+            out int fileSystemFlags,
+            char* fileSystemName,
+            int fileSystemNameBufLen)
+        {
+            // Allocate output buffer on the stack initially.
+            const int stackAllocation = 512;
+            Span<char> volumePathBuffer = wantsVolumePath ? stackalloc char[stackAllocation + 1] : stackalloc char[0]; // +1 to ensure a \0 at the end, todo: is this necessary
```

Comment:
> is this necessary

The docs say the max is MAX_PATH+1 for all the buffers this API is concerned with
https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getvolumeinformationbyhandlew

we have a define for that
https://github.com/dotnet/runtime/blob/0cf1f09d6a90d873e206a2357feafe5f05e56b43/src/libraries/Common/src/Interop/Windows/Kernel32/Interop.MAX_PATH.cs#L8

Reply:
What about for long path support? Would it potentially be longer with that?

Reply:
I'm just reading the docs. AFAIK long paths are only supported by a subset of IO API's. Maybe there's a list somewhere but I'd also expect it to not say MAX_PATH if that was the case. (@JeremyKuhne ?)

Reply:
@danmoseley What is being passed back is not a full path so the concept of long path doesn't apply for this particular API. For the most part all IO APIs that work with full paths support long paths as it is a fundamental part of the design of NT. You really have to get into the corners of Win32 to hit APIs that have hard-coded buffers and checks (I don't know of any concrete examples, but I'm sure they exist).

As you mentioned, not bothering to get the volume path in this case is more relevant. :)